### PR TITLE
many: fetch the device store assertion together and in the context of interpreting snap-declarations

### PR DESCRIPTION
**Pull request by pedronis**
Also for consistency make sure we are seeded and we have a model before allowing install/refresh operations, reporting otherwise a conflict error about being too early/in conflict with device setting up.

TODO:  (for follow up) make sure prepare-image also fetches the device store assertion if available

### CODE REVIEW

File: overlord/assertstate/assertmgr.go [hunk 1]
```diff
@@ -95,8 +96,29 @@ func doValidateSnap(t *state.Task, _ *tomb.Tomb) error {
...
+
+	err = doFetch(st, snapsup.UserID, func(f asserts.Fetcher) error {
+		if err := snapasserts.FetchSnapAssertions(f, sha3_384); err != nil {
+			return err
+		}
+
+		// fetch store assertion if available
+		if modelAs.Store() != "" {
+			err := snapasserts.FetchStore(f, modelAs.Store())
+			if notFound, ok := err.(*asserts.NotFoundError); ok {
```
Comment (mvo5) [inline overlord/assertstate/assertmgr.go]: This looks like we should have something like `snapasserts.IsNotFound(err)` helper (not a blocker for this PR though) which might make this easier to read but of course there is a complication with the notFound type so maybe not. Anyway, just a idle thought.

File: overlord/configstate/configcore/proxy.go [hunk 2]
```diff
@@ -106,9 +106,13 @@ func validateProxyStore(tr Conf) error {
...
 	st.Lock()
 	defer st.Unlock()
-	_, err = assertstate.Store(st, proxyStore)
+
+	store, err := assertstate.Store(st, proxyStore)
 	if asserts.IsNotFound(err) {
 		return fmt.Errorf("cannot set proxy.store to %q without a matching store assertion", proxyStore)
 	}
+	if err == nil && store.URL() == nil {
+		return fmt.Errorf("cannot set proxy.store to %q with a matching store assertion with url unset", proxyStore)
```
Comment (mvo5) [inline overlord/configstate/configcore/proxy.go]: (nitpick^2) The error message sounds a bit unnatural, but I'm not sure about a better one. Maybe: `cannot set proxy.store to %q when the matching store assertion has no URL field` or similar?

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -2054,3 +2070,36 @@ func GadgetConnections(st *state.State) ([]snap.GadgetConnection, error) {
...
+// hook setup by devicestate
+var (
+	Model func(st *state.State) (*asserts.Model, error)
+)
+
+// ModelPastSeed returns the device model assertion if available and
+// the device is seeded, at that point the device store is known
+// and seeding done. Otherwise it returns a ChangeConflictError
+// about being too early.
+func ModelPastSeed(st *state.State) (*asserts.Model, error) {
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go]: (nitpick^2) I wonder if we can find a more descriptive name. Some ideas: `ModelAfterSeeding`, `SeededModel`?

File: overlord/assertstate/assertmgr.go (hunk 1 above)
Comment (pedronis) [reply overlord/assertstate/assertmgr.go]: we do have the helper (I think is even used somewhere else in the PR)  asserts.IsNotFound but indeed in this case we care about at which level in the chain we didn't find something so we need the more complicated code

**Pull request merged**